### PR TITLE
Fix #4601 - add some useful log messages when requested WWR is rejected

### LINKED ISSUE #4601: setWindowToWallRatio failed silently without warnings
Issue overview
--------------

The setWindowToWallRatio function failed when specifying a WWR of 0.98 (98%) with **0.0** sill offset and **true** for heightOffsetFromFloor.

## Current Behavior

```ruby
m = Model.new
pts = OpenStudio::Point3dVector.new;  pts << OpenStudio::Point3d.new(0, 0, 0); pts << OpenStudio::Point3d.new(0, 1, 0); pts << OpenStudio::Point3d.new(1, 1, 0); pts << OpenStudio::Point3d.new(1, 0, 0);
s = Space::fromFloorPrint(pts, 3, m).get()
surface = m.getSurfaces.select{|s| s.surfaceType.upcase == 'WALL'}.first
new_window = surface.setWindowToWallRatio(0.98, 0.0, true).get

=> RuntimeError: Optional not initialized
```

## Expected Behavior

Expect: 1. window to wall ratio successfully set to 98% or above. or
2. Failed with informative messages.

## Steps to Reproduce

1. Create medium office prototype with OpenStudio Standard. Use '90.1-2019' as template. 
2. Load model
3. Loop through surfaces: model.getSurfaces.each do |ss|
4. get original construction of the subsurface
5. remove the subsurfaces in the surface
6. create new window: ss.setWindowToWallRatio(0.98, 0.0, true).get

### Environment
OpenStudio V3.2

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Windows 10
 - Version of OpenStudio (if using an intermediate build, include SHA): 3.2.1

### Context

Try to implement the logic in the [ASHRAE 90.1 2019 addendum l to 2019](https://www.ashrae.org/file%20library/technical%20resources/standards%20and%20guidelines/standards%20addenda/90_1_2019_l_20201030.pdf)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fix #4601

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/geometry/Geometry.cpp [hunk 1]
```diff
@@ -693,6 +698,12 @@ bool applyViewAndDaylightingGlassRatios(double viewGlassToWallRatio, double dayl
   double requestedTotalWindowArea = totalWWR * wallArea;
 
   if (requestedTotalWindowArea > maxWindowArea) {
+    LOG_FREE(
+      Warn, "utilities.geometry.applyViewAndDaylightingGlassRatios",
+      "Requested WWR ("
+        << totalWWR << ") is greater than the Max WWR (" << maxWindowArea / wallArea
+        << "). This limit is meant to preserve mandatory space between window and wall: 0.0254m (1 inch) on all sides + minViewToDaylightDistance ("
+        << minViewToDaylightDistance << "m).");
```
Comment (jmarrec) [inline src/utilities/geometry/Geometry.cpp:706]: New message for #4601.

File: src/utilities/geometry/Geometry.cpp [hunk 2]
```diff
@@ -674,10 +674,15 @@ bool applyViewAndDaylightingGlassRatios(double viewGlassToWallRatio, double dayl
...
   if (wallWidth < 2 * minGlassToEdgeDistance) {
+    LOG_FREE(Warn, "utilities.geometry.applyViewAndDaylightingGlassRatios",
+             "Wall Width (" << wallWidth << "m) is too small to preserve 0.0254m (1 inch) on both sides.");
     return false;
   }
 
   if (wallHeight < 2 * minGlassToEdgeDistance + minViewToDaylightDistance) {
+    LOG_FREE(Warn, "utilities.geometry.applyViewAndDaylightingGlassRatios",
+             "Wall Height (" << wallHeight << "m) is too small to preserve 0.0254m (1 inch) on both sides plus the minViewToDaylightDistance ("
+                             << minViewToDaylightDistance << "m).");
```
Comment (jmarrec) [inline src/utilities/geometry/Geometry.cpp:685]: Couple more

File: src/model/test/Surface_GTest.cpp [hunk 3]
```diff
@@ -1156,6 +1156,51 @@ TEST_F(ModelFixture, Surface_SouthWall_WWR) {
...
+
+  boost::optional<SubSurface> window = wall.setWindowToWallRatio(wwr, offset, true);
+  EXPECT_FALSE(window);
+  EXPECT_EQ(1, sink.logMessages().size());
+  EXPECT_EQ(Warn, sink.logMessages().front().logLevel());
+  EXPECT_EQ("utilities.geometry.applyViewAndDaylightingGlassRatios", sink.logMessages().front().logChannel());
+  EXPECT_EQ("Requested WWR (0.98) is greater than the Max WWR (0.900981). This limit is meant to preserve mandatory space between window and wall: "
+            "0.0254m (1 inch) on all sides + minViewToDaylightDistance (0m).",
+            sink.logMessages().front().logMessage());
+}
```
Comment (jmarrec) [inline src/model/test/Surface_GTest.cpp:1202]: New test

Comment (tijcolem) [reply src/model/test/Surface_GTest.cpp:1202]: nice! looks good

File: src/utilities/geometry/Geometry.cpp (hunk 1 above)
Comment (tijcolem) [reply src/utilities/geometry/Geometry.cpp:706]: perfect

**Review by tijcolem — APPROVED** [review]

**Pull request merged**